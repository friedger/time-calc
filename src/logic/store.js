import { createStore, applyMiddleware, compose } from 'redux'
import { createLogger } from 'redux-logger'
import rootReducer from './reducers/reducers'
import createSagaMiddleware from 'redux-saga'
import rootSaga from './sagas/middlewares'

const composeEnhancers = typeof window !== 'undefined' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ : compose
const sagaMiddleware = createSagaMiddleware()

let middlewares = [sagaMiddleware]

if (process && process.env.NODE_ENV !== 'test') {
  middlewares = [ ...middlewares, createLogger() ]
}

export default function configureStore (preloadedState) {
  const store = createStore(
    rootReducer,
    preloadedState,
    composeEnhancers(applyMiddleware(...middlewares))
  )
  sagaMiddleware.run(rootSaga)

  return store
}
