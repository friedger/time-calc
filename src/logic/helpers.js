import store from "store/dist/store.modern";
import Moment from "moment";
import "moment-duration-format";
import { parse } from "json2csv";

const blockstack = require("blockstack");

const STORE_CURRENT_PROJECT_ID = "currentProjectId";
const STORE_PROJECTS = "projects";
const STORE_TIMES = "times";

export class CalculationHelper {
  static fetchCalculation(form) {
    return CalculationHelper.calculateLocal(form);
  }

  static calculateLocal(form) {
    // eslint-disable-next-line no-console
    console.log(form);
    const breakDuration = Moment.duration(form.break);
    const startDate = new Moment(form.start, "HH:mm");
    const endDate = new Moment(form.end, "HH:mm");
    const milliseconds = endDate.subtract(breakDuration).diff(startDate);
    const duration = Moment.duration(milliseconds / 1000, "seconds").format(
      "HH:mm",
      { trim: false }
    );

    return {
      start: form.start,
      end: form.end,
      break: form.break,
      duration: duration,
      date: form.date,
      description: form.description,
      id: form.id,
      projectId: form.projectId
    };
  }
}

export class TimeHelper {
  static sortTimes(times) {
    return times.sort((a, b) => {
      a = Moment(a.date, "L");
      b = Moment(b.date, "L");

      if (a.isBefore(b)) {
        return -1;
      }

      if (a.isAfter(b)) {
        return 1;
      }
      return 0;
    });
  }

  static downloadTimes(times) {
    if (!times) {
      return;
    }

    const headers = Object.getOwnPropertyNames(times[0]).filter(
      name => name !== "index"
    );
    const csv = parse(times, { fields: headers });
    const csvContent = "data:text/csv;charset=utf-8;base64," + btoa(csv);

    window.open(csvContent);
  }

  static moment(value, format) {
    if (value instanceof Moment) {
      return value.format(format);
    }

    if (value) {
      return new Moment(value, format);
    }

    return new Moment();
  }

  static now() {
    return TimeHelper.time(new Moment());
  }

  static today() {
    return TimeHelper.date(new Moment());
  }

  static time(value) {
    return TimeHelper.moment(value, "HH:mm");
  }

  static date(value) {
    return TimeHelper.moment(value, "L");
  }
}

export class StoreHelper {
  static loadTimes() {
    return store.get(STORE_TIMES) || [];
  }

  static saveTimes(times) {
    store.set(STORE_TIMES, times);
  }
}

export class ProjectHelper {
  static loadCurrentProject() {
    let currentProjectId = store.get(STORE_CURRENT_PROJECT_ID) || undefined;
    let projects = ProjectHelper.loadProjects();
    let currentProject;

    if (currentProjectId) {
      currentProject = projects.find(p => p.id === currentProjectId);
    } else {
      currentProject = undefined;
    }

    if (!currentProject) {
      currentProject = ProjectHelper.createProject(projects);
    }

    return currentProject;
  }

  static createProject(projects, title) {
    let currentProject = {
      filename: "times.json",
      title: title || "Unnamed",
      id: uuid()
    };
    projects.push(currentProject);
    store.set(STORE_PROJECTS, projects);
    return currentProject;
  }

  static saveCurrentProject(project) {
    if (!project.id) {
      project.id = uuid();
    }
    if (!project.filesname) {
      project.filename = "times.json";
    }

    let projects = ProjectHelper.loadProjects();

    let index = projects.findIndex(p => p.id === project.id);
    if (index < 0) {
      projects.push(project);
    } else {
      projects[index] = Object.assign(projects[index], project); // keep other properties
    }
    store.set(STORE_PROJECTS, projects);

    let currentProjectId = store.get(STORE_CURRENT_PROJECT_ID) || undefined;
    if (currentProjectId !== project.id) {
      store.set(STORE_CURRENT_PROJECT_ID, project.id);
    }
  }

  static loadProjects() {
    return store.get(STORE_PROJECTS) || [];
  }

  static saveProjects(projects) {
    store.set(STORE_PROJECTS, projects);
  }
}

export class UserHelper {
  static signIn() {
    try {
      blockstack.redirectToSignIn(
        `${window.location.origin}/app`,
        `${window.location.origin}/manifest.json`,
        ["store_write", "publish_data"]
      );
    } catch (e) {
      // eslint-disable-next-line no-console
      console.log(e);
    }
  }

  static signOut() {
    blockstack.signUserOut();
  }

  static isSignInPending() {
    return blockstack.isSignInPending();
  }

  static handlePendingSignIn() {
    return blockstack.handlePendingSignIn();
  }

  static isUserSignedIn() {
    return blockstack.isUserSignedIn();
  }

  static loadUserData() {
    return blockstack.loadUserData();
  }
}

export class SyncHelper {
  static savePubKey() {
    const publicKey = blockstack.getPublicKeyFromPrivate(
      blockstack.loadUserData().appPrivateKey
    );
    blockstack
      .putFile("key.json", JSON.stringify(publicKey), { encrypt: false })
      .then(() => {})
      .catch(e => {
        // eslint-disable-next-line no-console
        console.log(e);
      });
  }

  static sync(filename) {
    return blockstack.putFile(
      filename,
      JSON.stringify(StoreHelper.loadTimes())
    );
  }

  static load(filename, username) {
    if (!username) {
      return blockstack.getFile(filename).then(
        function(timesString) {
          // eslint-disable-next-line no-console
          console.log("times " + timesString);
          if (timesString) {
            return JSON.parse(timesString).filter(t => t != null);
          } else {
            return null;
          }
        },
        function(error) {
          // eslint-disable-next-line no-console
          console.log("error load " + error);
          return [];
        }
      );
    } else {      
      const profile = blockstack.loadUserData();
      const options = { decrypt: false, username };
      const sharedFilename = `shared/${profile.username}/${filename}`;
      // eslint-disable-next-line no-console
      console.log("loading " + sharedFilename);
      return blockstack.getFile(sharedFilename, options).then(
        function(timesString) {
          // eslint-disable-next-line no-console
          console.log("times " + timesString);
          return JSON.parse(blockstack.decryptContent(timesString)).filter(
            t => t != null
          );
        },
        function(error) {
          // eslint-disable-next-line no-console
          console.log("error load " + error);
          return [];
        }
      );
    }
  }

  static syncProjectList() {
    blockstack.putFile("projects.json", ProjectHelper.loadProjects());
  }

  static allFiles() {
    const files = [];
    let profile = blockstack.loadUserData();

    return blockstack
      .getAppBucketUrl(profile.hubUrl, profile.appPrivateKey)
      .then(bucketUrl => {
        return blockstack
          .listFiles(f => {
            files.push(bucketUrl + f);
            return true;
          })
          .then(function() {
            return files;
          });
      });
  }

  static requestApproval(filename, username) {
    return () => {
      // eslint-disable-next-line no-console
      console.log("get key for " + username);
      const options = {
        decrypt: false,
        username,
        zoneFileLookupURL: "https://core.blockstack.org/v1/names"
      };
      try {
        return blockstack.getFile("key.json", options).then(
          file => {
            const publicKey = JSON.parse(file);
            const sharedFilename = `shared/${username}/${filename}`;
            // eslint-disable-next-line no-console
            console.log("shared file " + sharedFilename);
            return blockstack.putFile(
              sharedFilename,
              blockstack.encryptContent(
                JSON.stringify(StoreHelper.loadTimes()),
                { publicKey }
              ),
              { encrypt: false }
            );
          },
          error => {
            // eslint-disable-next-line no-console
            console.log("err" + error);
          }
        );
      } catch (e) {
        // eslint-disable-next-line no-console
        console.log("err" + e);
      }
    };
  }
}

export function uuid() {
  return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, function(c) {
    var r = (Math.random() * 16) | 0,
      v = c == "x" ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });
}
