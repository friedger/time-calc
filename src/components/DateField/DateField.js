import React from 'react'
import PropTypes from 'prop-types'

import DatePicker from 'material-ui-pickers/DatePicker'

import { TimeHelper } from '../../logic/helpers'

const dateMask = (value) => {
  const chars = value.split('')

  const months = [/[0-1]/, chars[0] === '0' ? /[0-9]/ : /[0-2]/]
  const days = [/[0-3]/, chars[3] === '3' ? /[0-1]/ : /[0-9]/]
  const years = [/[1-3]/, /[0-9]/, /[0-9]/, /[0-9]/]

  return months.concat('/').concat(days).concat('/').concat(years)
}

const DateField = ({input, label, showPicker}) => <DatePicker
  value={input.value ? TimeHelper.date(input.value) : null}
  onChange={(date) => input.onChange(TimeHelper.date(date))}
  keyboard={showPicker !== false}
  label={label}
  format='L'
  autoOk
  mask={dateMask}
  placeholder={TimeHelper.today()}
  disableOpenOnEnter
  fullWidth
  showTodayButton
  InputLabelProps={{
    shrink: true
  }}
/>

DateField.propTypes = {
  input: PropTypes.object,
  label: PropTypes.string,
  showPicker: PropTypes.bool
}

export default DateField
