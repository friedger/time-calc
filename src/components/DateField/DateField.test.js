import React from 'react'
import DateField from './DateField'
import { shallow } from 'enzyme'
import 'jest-enzyme'

import MomentUtils from 'material-ui-pickers/utils/moment-utils'
import MuiPickersUtilsProvider from 'material-ui-pickers/utils/MuiPickersUtilsProvider'
import moment from 'moment'
import { TimeHelper } from '../../logic/helpers'
import { IconButton } from '@material-ui/core'

describe('Component - Datefield', () => {
  const update = jest.fn()

  const validValue = {value: '05/25/2018', onChange: update}
  const invalidValue = {value: '05/32/2018', onChange: update}

  it('renders correctly', () => {
    const component = shallow(<MuiPickersUtilsProvider utils={MomentUtils} moment={moment}><DateField input={validValue} label='foo' /></MuiPickersUtilsProvider>)

    expect(component.html().replace(TimeHelper.today(), 'TODAY')).toMatchSnapshot()
  })

  it('can hide the picker', () => {
    const component = shallow(<MuiPickersUtilsProvider utils={MomentUtils} moment={moment}><DateField showPicker={false} input={validValue} label='foo' /></MuiPickersUtilsProvider>)

    expect(component.html().replace(TimeHelper.today(), 'TODAY')).toMatchSnapshot()
  })

  it('can render a default value', () => {
    const component = shallow(<MuiPickersUtilsProvider utils={MomentUtils} moment={moment}><DateField showPicker={false} input={{onChange: update}} label='foo' /></MuiPickersUtilsProvider>)

    expect(component.html().replace(TimeHelper.today(), 'TODAY')).toMatchSnapshot()
  })

  it('opens the datepicker on adornment click', () => {
    const component = shallow(<MuiPickersUtilsProvider utils={MomentUtils} moment={moment}><DateField input={validValue} label='foo' /></MuiPickersUtilsProvider>)

    component.find(IconButton).simulate('click')

    // expect(component.html().replace(TimeHelper.today(), "TODAY")).toMatchSnapshot();
  })

  xdescribe('Form Handling', () => {
    it('set the form value from props', () => {
      const component = shallow(<DateField formValue={validValue} select='date' />).dive()

      expect(component.find('input')).toHaveValue('bar')
    })

    xit('is always dirty', () => {
      const component = mount(<DateField formValue={validValue} select='date' />)

      component.update()
      expect(component.find('div')).toHaveClassName('is-dirty')
    })

    it('is marked invalid in case of errors', () => {
      let mounted = mount(<DateField formValue={invalidValue} select='date' />)
      expect(mounted.find('div').first()).toHaveClassName('is-invalid')

      mounted = mount(<DateField formValue={{errorList: [], value: undefined, update: update}} select='date' />)
      expect(mounted.find('div').first()).not.toHaveClassName('is-invalid')

      mounted.setProps({formValue: {errorList: [{foo: 'bar'}], value: undefined, update: update}})
      expect(mounted.find('div').first()).toHaveClassName('is-invalid')
    })

    it('invokes the form.update on value change', () => {
      const e = {target: {value: 'foo'}}

      const component = shallow(<DateField formValue={validValue} select='date' />).dive()

      component.find('input').simulate('change', e)
      expect(update).toBeCalledWith('foo')
      expect(component.state('dialog').time).toBeInstanceOf(Moment)
    })
  })
})
