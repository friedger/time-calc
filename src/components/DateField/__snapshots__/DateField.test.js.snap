// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Datefield can hide the picker 1`] = `" <div class=\\"MuiFormControl-root-2 MuiFormControl-fullWidth-5\\"><label class=\\"MuiFormLabel-root-11 MuiInputLabel-root-6 MuiInputLabel-formControl-7 MuiInputLabel-animated-10 MuiInputLabel-shrink-9\\" data-shrink=\\"true\\">foo</label><div class=\\"MuiInput-root-16 MuiInput-fullWidth-23 MuiInput-formControl-17 MuiInput-underline-20 WithUtils-DateTextField--input-1\\"><input type=\\"text\\" aria-invalid=\\"false\\" aria-required=\\"false\\" class=\\"MuiInput-input-24\\" placeholder=\\"TODAY\\" readonly=\\"\\" value=\\"05/25/2018\\"/></div></div> "`;

exports[`Component - Datefield can render a default value 1`] = `" <div class=\\"MuiFormControl-root-2 MuiFormControl-fullWidth-5\\"><label class=\\"MuiFormLabel-root-11 MuiInputLabel-root-6 MuiInputLabel-formControl-7 MuiInputLabel-animated-10 MuiInputLabel-shrink-9\\" data-shrink=\\"true\\">foo</label><div class=\\"MuiInput-root-16 MuiInput-fullWidth-23 MuiInput-formControl-17 MuiInput-underline-20 WithUtils-DateTextField--input-1\\"><input type=\\"text\\" aria-invalid=\\"false\\" aria-required=\\"false\\" class=\\"MuiInput-input-24\\" placeholder=\\"TODAY\\" readonly=\\"\\" value=\\"\\"/></div></div> "`;

exports[`Component - Datefield renders correctly 1`] = `" <div class=\\"MuiFormControl-root-2 MuiFormControl-fullWidth-5\\"><label class=\\"MuiFormLabel-root-11 MuiInputLabel-root-6 MuiInputLabel-formControl-7 MuiInputLabel-animated-10 MuiInputLabel-shrink-9\\" data-shrink=\\"true\\">foo</label><div class=\\"MuiInput-root-16 MuiInput-fullWidth-23 MuiInput-formControl-17 MuiInput-underline-20 WithUtils-DateTextField--input-1\\"><input type=\\"text\\" aria-invalid=\\"false\\" aria-required=\\"false\\" class=\\"MuiInput-input-24\\" placeholder=\\"TODAY\\"/><div class=\\"MuiInputAdornment-root-29 MuiInputAdornment-positionEnd-31\\"><button tabindex=\\"0\\" class=\\"MuiButtonBase-root-38 MuiIconButton-root-32\\" type=\\"button\\"><span class=\\"MuiIconButton-label-37\\"> <span class=\\"material-icons MuiIcon-root-41\\" aria-hidden=\\"true\\"> event </span> </span><span class=\\"MuiTouchRipple-root-47\\"></span></button></div></div></div> "`;
